### PR TITLE
Add basic zip handling from the CLI (fixes #27)

### LINKED ISSUE #27: Connect CLI to zip parser

### PR DESCRIPTION
**Pull request by muffinresearch**
* Hooks up zip handling to the cli. 
* Checks file exists
* Parses the zip.
* Adds colors + turns them off when --boring is passed.

Next up is starting to validate the layout + work out some of the structure for errors.

1__bash

### CODE REVIEW

File: src/cli.js [hunk 1]
```diff
@@ -24,7 +23,11 @@ export var cli = argv
     default: 'text',
     choices: ['json', 'text'],
   })
-  // self-hosted add-on.
+  .option('stack', {
```
Comment (tofumatt) [inline src/cli.js]: Nice :+1:

File: src/main.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import cli from 'cli';
+import Validator from 'validator';
+
+export function createInstance() {
+  return new Validator(cli.argv);
```
Comment (tofumatt) [inline src/main.js]: I like that API.

File: src/validator.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  constructor(config) {
+    this.config = config;
+    this.packagePath = config._[0];
+    this.xpi;
+    this.chalk = new chalk.constructor(
+      {enabled: !this.config.boring});
+  }
+
+  handleError(err) {
+    if (this.config.stack === true) {
```
Comment (tofumatt) [inline src/validator.js]: This is a nitpick, but testing for truthiness here would seem to suffice, right?

Would you prefer always using the explicit `=== true` going forward? If that's how you did things on payments/you prefer it that's cool, just making notes for #21.

File: src/validator.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      console.error(err.stack);
+    } else {
+      console.error(this.chalk.red(err.message || err));
+    }
+  }
+
+  checkFileExists(filepath, _lstat=lstat) {
+    var invalidMsg = new Error(
+      'Path given is not a file or does not exist: "' + filepath + '"');
```
Comment (tofumatt) [inline src/validator.js]: We should use template strings here instead, taking care of ES6 niceness. I think we did this elsewhere and I forgot to mention it, but:

```js
var invalidMessage = new Error(
  `Path given is not a file or does not exist: "${filepath}"`);
```
should work. Also, `invalidMessage` is a nicer variable name, we have space :smile:

File: tests/test.validator.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        assert.fail(null, null, 'Unexpected success');
+      })
+      .catch(err => {
+        assert.instanceOf(err, Error);
+        assert.include(err.message, 'Path given is not a file');
+      });
+  });
+
+  it('should detect other errors during lstat', () => {
+    var av = new Validator({_: ['foo']});
```
Comment (tofumatt) [inline tests/test.validator.js]: `av` tripped me up for a second, then I realised it meant `addonsValidator`. Is `validatorInstance` too verbose? I know `validator` is confusing/vague, but one/two letter variable names trip me up.

File: src/validator.js (hunk 4 above)
Comment (tofumatt) [inline src/validator.js]: And actually, reading the output I find this sentence hard to parse. This might read better:

```js
var invalidMessage = new Error(
  `Path "${filepath}" is not a file or does not exist.`);
```

This way you can see the path right away and the sentence is easier to mentally parse. I find that worth the trade-off of not seeing the path clearly at the end of the error (it's in quotes anyway, so it stands out).

**Comment by tofumatt** [comment]
r+wc

Looks great overall; see my few comments but they're mostly minor things.

File: src/validator.js (hunk 3 above)
Comment (muffinresearch) [inline src/validator.js]: I'm not sure if that's a thing we particularly adhered to in payments, but I definitely tend to default using explicit comparisons over truthy/falsey ones.

File: src/validator.js (hunk 4 above)
Comment (muffinresearch) [inline src/validator.js]: Good call.

File: tests/test.validator.js (hunk 5 above)
Comment (muffinresearch) [inline tests/test.validator.js]: Yeah in tests I might do something like this for brevity. Happy to be a bit more verbose though.

File: src/validator.js (hunk 3 above)
Comment (tofumatt) [inline src/validator.js]: Fair enough. I’ll just make a note in the JSCS so we keep things consistent; I tend to opt for truthy but a failed test for style will be a great reminder for me ;-)

- tofumatt

On 30 September 2015 at 09:20:17, Stuart Colville (notifications@github.com) wrote:

In src/validator.js:

> +
> +export var lstat = promisify(fs.lstat);
> +
> +
> +export default class Validator {
> +
> +  constructor(config) {
> +    this.config = config;
> +    this.packagePath = config._[0];
> +    this.xpi;
> +    this.chalk = new chalk.constructor(
> +      {enabled: !this.config.boring});
> +  }
> +
> +  handleError(err) {
> +    if (this.config.stack === true) {
I'm not sure if that's a thing we particularly adhered to in payments, but I definitely tend to default using explicit comparisons over truthy/falsey ones.
